### PR TITLE
fix/17: improve redirect loop detection performance

### LINKED ISSUE #17: check_for_possible_redirect_loops runs in Theta( n^2 ) time
I was using this plugin on a project with ~250 redirects. I wrote an import script that made use of the create_redirect method to generate the redirects. After creating the redirects, the WordPress backend was "frozen". It seemed as though something was caught in an infinite loop.

After some searching I realized the problem was the check_for_possible_redirect_loops(). This method checks each redirect against every other redirect meaning it runs in Theta( n^2 ) time where n is the number of redirects. With 250 redirects, the contents of the loop in that function will be executed over 62,000 times.

Proposal: Add another filter "max_redirects_to_check_for_loops" or something to that effect. It should default to 50.

### PR DESCRIPTION
**Pull request by Sidsector9**
Closes #17

## Description of the Change

This PR implements loop detection by first converting the redirect data to a directed graph, then performs a depth-first traversal to detect loops.

Below is a chart that describes the time complexity (worst case) between the new and the old implementation.
The worst case would be when there are no loops detected (no early exit) and the algorithm has to iterate through all the redirects.

The improvement is from `θ(n^2)` to `O(V+E)` where `V` and `E` are the number of nodes and edges representing a directed graph.

### Performance compared with `srm_check_for_possible_redirect_loops()`:

| Redirects | Before                    | After           |
|-----------|---------------------------|-----------------|
| 1000      | ~5.1 seconds              | ~0.00085 seconds |
| 10,000     | 560 seconds (~9 minutes) | ~2.5 seconds    |

### Both (old vs new) implementations

line-simple

In the chart above, the red graph grows quadratically, but the blue one grows linearly. Below, the zoomed-in version of the blue graph is shown.

### Performance for new implementation

After input size 500, the time complexity is almost linear as expected:

line-simple

## How to test the Change

1. Import this file - [srm-redirects.csv](https://github.com/10up/safe-redirect-manager/files/12427579/srm-redirects-v2.csv)  by running `wp safe-redirect-manager import srm-redirects.csv`. This file has around 2455 redirects.
2. Use the filter hook `srm_max_redirects` to return a value of `5000`.
3. Use the filter hook `srm_check_for_possible_redirect_loops` to return `true` to enable loop detection.
4. Delete transients to gauge performance without it using `wp transient delete --all`
5. Visit `/wp-admin/edit.php?post_type=redirect_rule`

To log time, you can use the following snippet:

```php
$old_start_time = microtime(true);
srm_check_for_possible_redirect_loops();
$old_end_time = microtime(true);
$old_execution_time = $old_end_time - $old_start_time;

$new_start_time = microtime(true);
SRM_Loop_Detection::detect_redirect_loops();
$new_end_time = microtime(true);
$new_execution_time = $new_end_time - $new_start_time;
error_log( number_format( $old_execution_time, 8 ) . ' > ' . number_format( $new_execution_time, 8 ) );
```

## Changelog Entry

> Changed - improve redirect loop detection performance

## Credits

Props @tlovett1 @danielbachhuber @Sidsector9 @ravinderk @iamdharmesh 

## Checklist:

- [ ] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

### CODE REVIEW

File: inc/classes/class-srm-post-type.php [hunk 1]
```diff
@@ -235,7 +235,7 @@ public function action_redirect_chain_alert() {
 			 * will be running slow. Let's disable it by default.
 			 */
 			if ( apply_filters( 'srm_check_for_possible_redirect_loops', false ) ) {
-				if ( srm_check_for_possible_redirect_loops() ) {
+				if ( SRM_Loop_Detection::detect_redirect_loops() ) {
 					?>
 					<div class="updated">
```
Comment (ravinderk) [inline inc/classes/class-srm-post-type.php]: @Sidsector9 Can you make this warning notice instead of a success notice?

**Review by ravinderk — DISMISSED** [review]
@Sidsector9 This is working as expected 👍 

I added a minor request that is not a blocker.

------------------------

After reviewing this pull request, we should not allow saving or importing redirects, which causes an infinite redirect. If we follow this, we do not need logic to check for redirect conflict.

**Review by iamdharmesh — APPROVED** [review]
**Note**: Not reviewed full PR, only reviewed the changes requested in PR feedback by @ravinderk

**Comment by Sidsector9** [comment]
Updated the PR with detailed analysis.

**Pull request merged**